sherpa/astro/ui/serialize.py

```diff
@@ -984,65 +1111,179 @@ def _save_models(state, fh=None):
                     # only one to support backgrounds
                     if have_full_model:
                         if repr(the_bkg_source) == repr(the_bkg_full_model):
-                            cmd = "set_bkg_full_model(%s, %s, bkg_id=%s)" % (
-                                cmd_id, the_bkg_full_model.name, cmd_bkg_id)
+                            cmd = f"set_bkg_full_model({cmd_id}, {the_bkg_full_model.name}, bkg_id={cmd_bkg_id})"
                         else:
-                            cmd = "set_bkg_source(%s, %s, bkg_id=%s)" % (
-                                cmd_id, the_bkg_source.name, cmd_bkg_id)
+                            cmd = f"set_bkg_source({cmd_id}, {the_bkg_source.name}, bkg_id={cmd_bkg_id})"
                     else:
-                        cmd = "set_bkg_source(%s, %s, bkg_id=%s)" % (
-                            cmd_id, the_bkg_source.name, cmd_bkg_id)
+                        cmd = f"set_bkg_source({cmd_id}, {the_bkg_source.name}, bkg_id={cmd_bkg_id})"
 
                 elif have_full_model:
-                    cmd = "set_bkg_full_model(%s, %s, bkg_id=%s)" % (
-                        cmd_id, the_bkg_full_model.name, cmd_bkg_id)
+                    cmd = f"set_bkg_full_model({cmd_id}, {the_bkg_full_model.name}, bkg_id={cmd_bkg_id})"
 
                 else:
-                    cmd = ""
+                    continue
 
-                _output(cmd, fh)
-                _output("", fh)
+                _output(out, cmd)
+                _output_nl(out)
 
         except:
             pass
 
+    # separate out the pileup models from the source models
+    #
+    for id in ids:
+        try:
+            pname = state.get_pileup_model(id).name
+        except:
+            continue
+
+        cmd_id = _id_to_str(id)
+        cmd = f"set_pileup_model({cmd_id}, {pname})"
+        _output(out, cmd)
+
+    # Do want to remove the initial banner?
+    #
+    _remove_banner(out, orig_pos)
+
 
-def _save_xspec(fh=None):
-    """Save the XSPEC settings, if the module is loaded.
+def _save_xspec(out: OutType) -> None:
+    """Save the XSPEC settings.
 
     Parameters
     ----------
-    fh : None or file-like
-       If ``None``, the information is printed to standard output,
-       otherwise the information is added to the file handle.
+    out : dict
+       The output state
     """
 
-    if not hasattr(sherpa.astro, "xspec"):
+    # This case shpuld not happen, but just in case.
+    #
+    if xspec is None:
         return
 
```

Comment:
or maybe?
```suggestion
    # This case shpuld not happen, but just in case.
    assert xspec is not None
```
potentially more disrupting to users if this case happens, but for us, we would know we've run into a problem.

Reply:
For this use case I don't want the user to be disrupted in case something strange has happened.